src/streamlit_app.py

```diff
@@ -382,106 +383,114 @@ def render_grouped_bar_chart(
 
-def render_chart(df: pd.DataFrame) -> bool:
-    # pylint: disable=too-many-branches
+
+def render_chart(df: pd.DataFrame, json_query: typing.Dict[str, typing.Any]) -> bool:
+    # pylint: disable=too-many-branches,too-many-return-statements
 
     if len(df) == 0:
         return False
 
+    metrics = _get_columns_and_types(json_query, "metrics")
+    attributes = _get_columns_and_types(json_query, "attributes")
+
+    if len(metrics) == 0:
+        # No chart when there are no metrics to show
+        return False
+
     # Bug in streamlit with dots in column names - update column names
-    updated_columns = {col: col.replace(".", "_") for col in df.columns}
+    updated_columns = {col: _fix_column_name_for_chart(col) for col in df.columns}
     df = df.rename(columns=updated_columns)
-    numeric_columns = list(df.select_dtypes(include=["number"]).columns)
 
-    if len(numeric_columns) == 0:
-        st.markdown("Chart currently not available for non-numeric data")
+    if set(metrics.values()) - {"number", "float"}:
+        st.markdown("Chart currently not available for non-numeric metrics")
         return False
 
     # Bug in snowflake connector - does not set date types correctly in pandas, so manually detecting
-    date_columns = get_possible_date_columns(df)
-    str_columns = df.select_dtypes(include=["object"]).columns
-    str_columns = [col for col in str_columns if col not in date_columns]
-    df_to_show = df[numeric_columns + str_columns]
-
-    for col in date_columns[:1]:
+    date_columns = [
+        col for col, col_type in attributes.items() if col_type in ("date", "timestamp")
+    ]
+    dim_columns = [col for col in attributes if col not in date_columns]
+    metric_columns = list(metrics.keys())
+    df_to_show = df[metric_columns]
+
+    for col in date_columns:
+        # Fix datatype - as Snowflake returns dates as objects
         df_to_show[col] = pd.to_datetime(df[col], errors="coerce")
 
-    # date
+    for col in dim_columns:
+        # Show all dimensions as strings
+        df_to_show[col] = df[col].astype(str)
+
+    # date dimension
     if len(date_columns) > 0:
         df_to_show = df_to_show.rename(columns={date_columns[0]: "index"}).set_index(
             "index",
         )
 
-        if len(str_columns) == 0:
+        if len(dim_columns) == 0:
             st.line_chart(df_to_show)
-        elif len(str_columns) == 1 and len(numeric_columns) == 1:
-            st.line_chart(df_to_show, y=numeric_columns[0], color=str_columns[0])
+        elif len(dim_columns) == 1 and len(metric_columns) == 1:
+            st.line_chart(df_to_show, y=metric_columns[0], color=dim_columns[0])
+        else:
+            return False
 
-    # text
-    elif len(str_columns) >= 1:
-        metric_column = numeric_columns[0]  # First numeric column
+    # dimensions
+    elif len(dim_columns) >= 1:
+        metric_column = metric_columns[0]  # First numeric column
 
         if len(df_to_show) > 50:
```

Comment:
Shouldn't we apply this in all cases? Now it's not covered in the date columns flow

Reply:
no as it handles very well compression of time series

Reply:
and sorting there would be wrong